src/explorers/glob_explorer.rs

```diff
@@ -1,12 +1,146 @@
+    for c in pattern.chars() {
+        if escaped {
+            escaped = false;
+
+            if depth == 0 {
+                tokens.push(ExtendGlobToken::Literal(c));
+            } else {
+                push_subpattern_character(&mut tokens, c);
+            }
+
+            continue;
+        }
+
+        match c {
+            '\\' => {
+                escaped = true;
+
+                if depth == 0 {
+                    tokens.push(ExtendGlobToken::Literal(c));
+                } else {
+                    push_subpattern_character(&mut tokens, c);
+                }
+            }
+            '{' => {
+                depth += 1;
+
+                if depth == 1 {
+                    // prepare for subpatterns at depth 1
+                    tokens.push(ExtendGlobToken::Subpatterns(vec!["".to_owned()]));
+                } else {
+                    push_subpattern_character(&mut tokens, c);
+                }
+            }
+            '}' => {
+                depth -= 1;
+
+                if depth == 0 {
+                    // closing the subpattern at depth 1: extend subpatterns recursively
+                    let mut extended_basic_glob_patterns: Vec<String> = Vec::new();
+
+                    match tokens.pop().unwrap() {
+                        ExtendGlobToken::Subpatterns(subpatterns) => {
+                            for subpattern in subpatterns {
+                                extended_basic_glob_patterns
+                                    .extend(extend_glob_pattern(&subpattern));
+                            }
+                        }
+                        ExtendGlobToken::Literal(_) => panic!("Invalid state"),
+                    }
+
+                    tokens.push(ExtendGlobToken::Subpatterns(extended_basic_glob_patterns));
+                } else {
+                    push_subpattern_character(&mut tokens, c);
+                }
+            }
+            ',' => {
+                if depth == 0 {
+                    tokens.push(ExtendGlobToken::Literal(c));
+                } else if depth == 1 {
+                    // delimits two subpatterns in the depth 1 disjuction; prepare for the next subpattern
+                    match tokens.last_mut().unwrap() {
+                        ExtendGlobToken::Subpatterns(subpatterns) => {
+                            subpatterns.push("".to_owned());
+                        }
+                        _ => panic!("Invalid state"),
+                    }
+                } else {
+                    push_subpattern_character(&mut tokens, c);
+                }
+            }
+            _ => {
+                if depth == 0 {
+                    tokens.push(ExtendGlobToken::Literal(c));
+                } else {
+                    push_subpattern_character(&mut tokens, c);
+                }
+            }
+        }
+    }
+
+    // reconstruct the basic glob patterns from the tokens; this is basically the cartesian product
```

Comment:
This should probably be refactored so that:
1. All the extend stuff is in a namespace
2. Extend is broken up into explode and cartesian product operations